system/Database/Query.php

```diff
@@ -297,15 +297,27 @@ public function getErrorMessage(): string
 
 	//--------------------------------------------------------------------
 
+	/**
+	 * Determines if the statement is a write-type query or not.
+	 *
+	 * @param string $sql
+	 * @return boolean
+	 */
+	public static function sqlIsWriteType($sql): bool
+	{
+		return (bool) preg_match('/^\s*"?(SET|INSERT|UPDATE|DELETE|REPLACE|CREATE|DROP|TRUNCATE|LOAD|COPY|ALTER|RENAME|GRANT|REVOKE|LOCK|UNLOCK|REINDEX|OPTIMIZE|MERGE)\s/i', $sql);
+	}
+
+	//--------------------------------------------------------------------
+
 	/**
 	 * Determines if the statement is a write-type query or not.
 	 *
 	 * @return boolean
 	 */
 	public function isWriteType(): bool
 	{
-		return (bool) preg_match(
-						'/^\s*"?(SET|INSERT|UPDATE|DELETE|REPLACE|CREATE|DROP|TRUNCATE|LOAD|COPY|ALTER|RENAME|GRANT|REVOKE|LOCK|UNLOCK|REINDEX)\s/i', $this->originalQueryString);
+		return self::sqlIsWriteType($this->originalQueryString);
 	}
```

Comment:
It is better to delegate the logic to the DB driver.
Because SQL is DBMS specific, so we should not make it be common.

Reply:
As I posted before the CI3 codebase has an [is_write_type function in the base driver class](https://github.com/bcit-ci/CodeIgniter/blob/18718545e0faa98d62df3f65d045d84494ec29ca/system/database/DB_driver.php) and only overrides it in these places:
* [odbc driver](https://github.com/bcit-ci/CodeIgniter/blob/18718545e0faa98d62df3f65d045d84494ec29ca/system/database/drivers/odbc/odbc_driver.php#L299)
* [PDO odbc driver](https://github.com/bcit-ci/CodeIgniter/blob/18718545e0faa98d62df3f65d045d84494ec29ca/system/database/drivers/pdo/subdrivers/pdo_odbc_driver.php#L182)
* [PDO pgsql driver](https://github.com/bcit-ci/CodeIgniter/blob/18718545e0faa98d62df3f65d045d84494ec29ca/system/database/drivers/pdo/subdrivers/pdo_pgsql_driver.php)
* [pgsql driver](https://github.com/bcit-ci/CodeIgniter/blob/18718545e0faa98d62df3f65d045d84494ec29ca/system/database/drivers/postgre/postgre_driver.php)

So we would need to override the baseline function I have now only in the case of Postgres for CI4.

Reply:
As you suggested, the latest changes have moved `isWriteType($sql)` into the BaseConnection class, and it is in some cases overridden by the DMBS-specific Connection classes (see Postgre\Connection and SQLSRV\Connection). The Query classes also have `isWriteType()` functions now, and these use the Connection objects passed into their constructors.